tensorflow/python/ops/rnn_cell_impl.py

```diff
@@ -1036,23 +1042,35 @@ def _should_dropout(p):
       return (not isinstance(p, float)) or p < 1
 
     if _should_dropout(self._input_keep_prob):
-      inputs = self._dropout(inputs, "input",
+      def dropped_inputs():
+        return self._dropout(inputs, "input",
                              self._recurrent_input_noise,
                              self._input_keep_prob)
+      inputs = utils.smart_cond(self._training,
+                                dropped_inputs,
+                                lambda: array_ops.identity(inputs))
     output, new_state = self._cell(inputs, state, scope=scope)
     if _should_dropout(self._state_keep_prob):
-      # Identify which subsets of the state to perform dropout on and
-      # which ones to keep.
-      shallow_filtered_substructure = nest.get_traverse_shallow_structure(
-          self._dropout_state_filter, new_state)
-      new_state = self._dropout(new_state, "state",
-                                self._recurrent_state_noise,
-                                self._state_keep_prob,
-                                shallow_filtered_substructure)
+      def dropped_new_state():
+        # Identify which subsets of the state to perform dropout on and
+        # which ones to keep.
+        shallow_filtered_substructure = nest.get_traverse_shallow_structure(
+            self._dropout_state_filter, new_state)
+        return self._dropout(new_state, "state",
+                             self._recurrent_state_noise,
+                             self._state_keep_prob,
+                             shallow_filtered_substructure)
+      new_state = utils.smart_cond(self._training,
+                                   dropped_new_state,
+                                   lambda: array_ops.identity(new_state))
```

Comment:
do you need the `array_ops.identity` ?

Reply:
In `tf.layers.dropout`, it applies `tf.identity` to `inputs` before returning, but I haven't able to fully understand why it has to be this way. I think (guess) it is to make sure the `input` tensor to be in the same device as the `dropped input (if applied)`.

In here, If I am guessing above correctly, `tf.identity` can be skipped as the `scope` or `device` etc. is decided only when the Cell object is applied?

I think `tf.identity` is a little bit difficult to understand, please correct me if I am wrong, thanks.

Reply:
@ebrevdo any comment is very appreciated.